### PR TITLE
Support by-floor XRay generation.

### PR DESCRIPTION
**Pull request by SirVer**
This changes the pipeline to no longer be a singleton and simplifies the interface for registering PointsProcessors.

### CODE REVIEW

File: cartographer/io/points_processor_pipeline_builder.cc [hunk 1]
```diff
@@ -26,34 +26,52 @@
...
-  RegisterNonStatic<CountingPointsProcessor>();
-  RegisterNonStatic<FixedRatioSamplingPointsProcessor>();
-  RegisterNonStatic<MinMaxRangeFiteringPointsProcessor>();
-  RegisterNonStatic<OutlierRemovingPointsProcessor>();
-  RegisterNonStatic<PcdWritingPointsProcessor>();
-  RegisterNonStatic<PlyWritingPointsProcessor>();
-  RegisterNonStatic<XRayPointsProcessor>();
-  RegisterNonStatic<XyzWriterPointsProcessor>();
+template <typename PointsProcessorType>
+void RegisterPlainPointsProcessor(PointsProcessorPipelineBuilder* builder) {
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.cc]: const

File: cartographer/io/points_processor_pipeline_builder.h [hunk 2]
```diff
@@ -23,53 +23,45 @@
...
 
 namespace cartographer {
 namespace io {
 
-// Singleton that knows how to build a points processor pipeline out of a Lua
-// configuration. All the PointsProcessor shipping with Cartographer are already
-// registered with 'instance', but can register new classes with it that must
-// define its name and a way to build itself out of a LuaParameterDictionary.
-// See the various 'PointsProcessor's for examples.
+// Builder to create a points processor pipeline out of a Lua configuration. You
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.h]: Builder to create a points processor pipeline out of a Lua configuration.

You can register all built-in PointsProcessors using 'RegisterBuiltInPointsProcessors'. Non-built-in PointsProcessors must define a name and a factory method for building itself from a LuaParameterDictionary. See the various built-in PointsProcessors for examples.

File: cartographer/io/points_processor_pipeline_builder.cc [hunk 3]
```diff
@@ -26,34 +26,52 @@
...
+      [](common::LuaParameterDictionary* const dictionary,
+         PointsProcessor* const next) -> std::unique_ptr<PointsProcessor> {
+        return PointsProcessorType::FromDictionary(dictionary, next);
+      });
 }
 
-PointsProcessorPipelineBuilder* PointsProcessorPipelineBuilder::instance() {
-  static PointsProcessorPipelineBuilder instance;
-  return &instance;
+void RegisterCartographerPointsProcessors(
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.cc]: s/Cartographer/BuiltIn

**Comment by SirVer** [comment]
ptal

**Pull request merged**